src/database.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Jenkins agent relation observer."""
+import logging
+import typing
+
+import psycopg2
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+from ops.charm import CharmBase
+from ops.framework import Object
+from psycopg2 import sql
+from psycopg2.extensions import connection
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseObserver(Object):
+    """The Database relation observer."""
+
+    _RELATION_NAME = "database"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        # SUPERUSER is required to update pg_database
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+            database_name=self._charm.app.name,
+            extra_user_roles="SUPERUSER",
+        )
+
+    def get_relation_data(self) -> typing.Optional[typing.Dict]:
+        """Get database data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+        """
+        if self.model.get_relation(self._RELATION_NAME) is None:
+            return None
+
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        endpoint = relation_data.get("endpoints", ":")
+
+        return {
+            "POSTGRES_USER": relation_data.get("username", ""),
+            "POSTGRES_PASSWORD": relation_data.get("password", ""),
+            "POSTGRES_HOST": endpoint.split(":")[0],
+            "POSTGRES_PORT": endpoint.split(":")[1],
+            "POSTGRES_DB": self._charm.app.name,
+        }
+
+    def prepare_database(self) -> None:
+        """Change database collate and ctype as required by Synapse.
+
+        Raises:
+           Error: something went wrong while preparing the database.
+        """
+        conn = self.get_conn()
+        database_name = self.get_database_name()
+        if conn is not None:
+            try:
+                with conn.cursor() as curs:
+                    curs.execute(
+                        "UPDATE pg_database "
+                        "SET datcollate='C', datctype='C' "
+                        "WHERE datname=%s",
```

Comment:
Is this vulnerable to a SQL injection attack? I think usually these kind of things are handled as variables in SQL rather than injecting directly into the SQL statement

Reply:
Also, What is the reason for picking `%s` compared to f-strings?

Reply:
Because this is the way recommended by the library documentation in "[Passing parameters to SQL queries](https://www.psycopg.org/docs/usage.html#query-parameters)".

I found out this because I tried to use f-string and this was the result:
```>> Issue: [B608:hardcoded_sql_expressions] Possible SQL injection vector through string-based query construction.
  Severity: Medium   Confidence: Medium
  CWE: CWE-89 (https://cwe.mitre.org/data/definitions/89.html)
  More Info: https://bandit.readthedocs.io/en/1.7.5/plugins/b608_hardcoded_sql_expressions.html
  Location: /home/ubuntu/github-runner/_work/synapse-operator/synapse-operator/src/database.py:78:24
77                        curs.execute(
78                            "UPDATE pg_database SET datcollate='C', datctype='C' "
79                            f"WHERE datname='{database_name}' AND datcollate='en_US.UTF-8' "
80                            "AND datctype='en_US.UTF-8';"
81                        )
```

Reply:
I'll change this in favour of sql.SQL and sql.Identifier.

Reply:
Changed.